test/test_lazy.py

```diff
@@ -0,0 +1,231 @@
+        dd = u_tpair.dd
+        dd_allfaces = dd.with_dtag("all_faces")
+        normal = thaw(discr.normal(dd), u_tpair.int[0].array_context)
+        flux = u_tpair.avg @ normal
+        return discr.project(dd, dd_allfaces, flux)
+
+    def op(u):
+        return div_operator(discr, u, get_flux(interior_trace_pair(discr, u)))
+
+    lazy_op = lazy_actx.compile(op)
+
+    def get_inputs(actx):
+        nodes = thaw(discr.nodes(), actx)
+        u = make_obj_array([actx.np.sin(np.pi*nodes[i]) for i in range(2)])
+        return u,
+
+    rel_linf_error = partial(_rel_linf_error, discr)
+
+    def lazy_to_eager(u):
+        return thaw(freeze(u, lazy_actx), eager_actx)
+
+    eager_result = op(*get_inputs(eager_actx))
+    lazy_result = lazy_to_eager(lazy_op(*get_inputs(lazy_actx)))
+    assert rel_linf_error(lazy_result, eager_result) < 1e-12
+
+
+@pytest.mark.parametrize("order", [1, 2, 3])
+def test_lazy_op_diffusion(op_test_data, order):
+    eager_actx, lazy_actx, get_discr = op_test_data
+    discr = get_discr(order)
+
+    from grudge.dof_desc import DTAG_BOUNDARY, DISCR_TAG_BASE
+    from mirgecom.diffusion import (
+        diffusion_operator,
+        DirichletDiffusionBoundary,
+        NeumannDiffusionBoundary)
+
+    boundaries = {
+        DTAG_BOUNDARY("x"): DirichletDiffusionBoundary(0),
+        DTAG_BOUNDARY("y"): NeumannDiffusionBoundary(0)
+    }
+
+    def op(alpha, u):
+        return diffusion_operator(
+            discr, DISCR_TAG_BASE, alpha, boundaries, u)
+
+    lazy_op = lazy_actx.compile(op)
+
+    def get_inputs(actx):
+        nodes = thaw(discr.nodes(), actx)
+        alpha = discr.zeros(actx) + 1
+        u = actx.np.cos(np.pi*nodes[0])
+        return alpha, u
+
+    rel_linf_error = partial(_rel_linf_error, discr)
+
+    def lazy_to_eager(u):
+        return thaw(freeze(u, lazy_actx), eager_actx)
+
+    eager_result = op(*get_inputs(eager_actx))
+    lazy_result = lazy_to_eager(lazy_op(*get_inputs(lazy_actx)))
+    assert rel_linf_error(lazy_result, eager_result) < 1e-12
+
+
+@pytest.mark.parametrize("order", [1, 2, 3])
+def test_lazy_op_euler(op_test_data, order):
+    eager_actx, lazy_actx, get_discr = op_test_data
+    discr = get_discr(order)
+
+    from grudge.dof_desc import DTAG_BOUNDARY
+    from mirgecom.eos import IdealSingleGas
+    from mirgecom.boundary import AdiabaticSlipBoundary
+    from mirgecom.euler import euler_operator
+
+    eos = IdealSingleGas()
+
+    boundaries = {
+        DTAG_BOUNDARY("x"): AdiabaticSlipBoundary(),
+        DTAG_BOUNDARY("y"): AdiabaticSlipBoundary()
+    }
+
+    def op(state):
+        return euler_operator(discr, eos, boundaries, state)
+
+    lazy_op = lazy_actx.compile(op)
+
+    def get_inputs(actx):
+        nodes = thaw(discr.nodes(), actx)
+        from mirgecom.initializers import MulticomponentLump
+        init = MulticomponentLump(
+            dim=2, velocity=np.ones(2), spec_y0s=np.ones(3),
+            spec_amplitudes=np.ones(3))
```

Comment:
This should produce mostly zeros in the RHS.  Only the passive advected-scalar components will get non-zero RHS.  Given how the relative error function above works, I'm wondering whether it would be prudent to add tests that test equation-at-a-time, for each equation.

Reply:
Is this case OK with the per-component error? Or is there maybe a better one I should use?

Reply:
The expected RHS for this case goes something like [0, 0 , 0 , 0, a, b, c].  

I'm not saying that just because the expected result is 0, that the equation is not exercised, but I am concerned that when the expected values are zero/small that we run right up against situation that can allow errors to escape undetected; when a potentially large point-local error is diminished by the max value of the field across the entire domain to which it is compared when calculating the relative error as it is written above. Dealing with this type of issue was important for fluid solution comparisons we did in pc2compare, for example.

I can't think of a (physical) case that obviously avoids this issue for all the equations - but I think we could come up with contrived cases for each equation that would give us more confidence in the results. 

FWIW, I think we should write a centralized solution comparison utility in the *MIRGE-Com* library. It should report errors to the user on a component-by-component basis, and revert to absolute differences when the relative comparisons will yield anomalous results.  I know that may seem like overkill for this particular application - but such a utility would be pretty useful in this and many other cases where our current error evaluations are, imo, falling way short.

Reply:
> The expected RHS for this case goes something like [0, 0 , 0 , 0, a, b, c].

Hm, I'm getting a nonzero RHS for all of the components for some reason:

```python
eager_components = eager_result.join()
print([discr.norm(comp) for comp in eager_components])
```

```
[9.797958971132713, 44.090815370097204, 6.9282032302755026, 6.928203230275506, 16.98354190507502, 16.98354190507502, 16.98354190507502]
```

...though I just noticed by BCs are inconsistent with me setting `velocity=np.ones(2)`, so I might have to tweak things.
 
> I can't think of a (physical) case that obviously avoids this issue for all the equations - but I think we could come up with contrived cases for each equation that would give us more confidence in the results.

We can always do multiple tests. I think we need that anyway, since we probably want to check that all of the boundary conditions work with lazy?

> FWIW, I think we should write a centralized solution comparison utility in the _MIRGE-Com_ library. It should report errors to the user on a component-by-component basis, and revert to absolute differences when the relative comparisons will yield anomalous results. I know that may seem like overkill for this particular application - but such a utility would be pretty useful in this and many other cases where our current error evaluations are, imo, falling way short.

👍

Reply:
> ```
> [9.797958971132713, 44.090815370097204, 6.9282032302755026, 6.928203230275506, 16.98354190507502, 16.98354190507502, 16.98354190507502]
> ```

Ho! 😕 😲 :concerned: